File: src/main/java/com/leijendary/spring/microservicetemplate/util/RequestContext.java
package com.leijendary.spring.microservicetemplate.util;

import org.springframework.lang.Nullable;
import org.springframework.web.context.request.ServletRequestAttributes;
import org.springframework.web.servlet.support.RequestContextUtils;

import javax.servlet.http.HttpServletRequest;
import java.net.URI;
import java.time.OffsetDateTime;
import java.util.Locale;
import java.util.TimeZone;

import static java.net.URI.create;
import static java.util.Optional.ofNullable;
import static org.springframework.security.core.context.SecurityContextHolder.getContext;
import static org.springframework.web.context.request.RequestContextHolder.getRequestAttributes;

public class RequestContext {

    @Nullable
    public static HttpServletRequest getCurrentRequest() {
        final var attributes = getRequestAttributes();

        if (!(attributes instanceof ServletRequestAttributes)) {
            return null;
        }

        return ((ServletRequestAttributes) attributes).getRequest();
    }

    @Nullable
    public static String getUsername() {
        return getContext().getAuthentication().getName();
    }

    @Nullable
    public static String getPath() {
        final var request = getCurrentRequest();

        if (request == null) {
            return null;
        }

        final var contextPath = request.getContextPath();

        return request.getRequestURI().replaceFirst(contextPath, "");
    }

    @Nullable
    public static URI uri() {
        final var request = getCurrentRequest();
        var path = getPath();

        if (request == null || path == null) {
            return null;
        }

        final var params = request.getQueryString();

        if (ofNullable(params).isPresent()) {
            path += "?" + params;
        }

        return create(path);
    }

    public static TimeZone getTimeZone() {
        return ofNullable(getCurrentRequest())
                .map(RequestContextUtils::getTimeZone)
                .orElse(TimeZone.getDefault());
    }

    public static Locale getLocale() {
        return ofNullable(getCurrentRequest())
                .map(RequestContextUtils::getLocale)
                .orElse(Locale.getDefault());
    }

    public static String getLanguage() {
        return getLocale().getLanguage();
    }

    public static OffsetDateTime now() {
        return OffsetDateTime.now(getTimeZone().toZoneId());
    }
}

File: src/main/kotlin/com/leijendary/spring/template/core/util/RequestContext.kt
package com.leijendary.spring.template.core.util

import com.leijendary.spring.template.core.config.HEADER_USER_ID
import com.leijendary.spring.template.core.config.properties.AuthProperties
import com.leijendary.spring.template.core.util.SpringContext.Companion.getBean
import jakarta.servlet.http.HttpServletRequest
import org.springframework.context.i18n.LocaleContextHolder.getLocale
import org.springframework.context.i18n.LocaleContextHolder.getTimeZone
import org.springframework.web.context.request.RequestContextHolder.getRequestAttributes
import org.springframework.web.context.request.ServletRequestAttributes
import java.net.URI
import java.time.OffsetDateTime
import java.time.ZoneId
import java.util.*

private val authProperties = getBean(AuthProperties::class)

object RequestContext {
    val currentRequest: HttpServletRequest?
        get() {
            val attributes = getRequestAttributes() as? ServletRequestAttributes

            return attributes?.request
        }

    val userId: String
        get() {
            return currentRequest
                ?.getHeader(HEADER_USER_ID)
                ?: authProperties.system.principal
        }

    val uri: URI?
        get() {
            val request = currentRequest ?: return null
            val contextPath = request.contextPath
            var url = request.requestURI.replaceFirst(contextPath.toRegex(), "")

            request.queryString?.also { url += "?$it" }

            return URI.create(url)
        }

    val timeZone: TimeZone
        get() = getTimeZone()

    val zoneId: ZoneId
        get() = timeZone.toZoneId()

    val locale: Locale
        get() = getLocale()

    val language: String
        get() = locale.language

    val now: OffsetDateTime
        get() = OffsetDateTime.now(zoneId)
}
